### PR TITLE
uefi: Implement getcwd and chdir

### PR DESCRIPTION
**Pull request by Ayush1325**
- Using EFI Shell Protocol. These functions do not make much sense unless a shell is present.
- Return the exe dir in case shell protocol is missing.

r? @joboet

### CODE REVIEW

File: library/std/src/sys/pal/uefi/os.rs [hunk 1]
```diff
@@ -125,11 +125,32 @@ pub fn error_string(errno: RawOsError) -> String {
...
+            // executables
+            assert!(t.pop());
+            Ok(t)
+        }
+    }
 }
 
-pub fn chdir(_: &path::Path) -> io::Result<()> {
-    unsupported()
+pub fn chdir(p: &path::Path) -> io::Result<()> {
```
Comment (bjorn3) [inline library/std/src/sys/pal/uefi/os.rs:146]: Chdir is supposed to only affect the current process and children, not it's parent.

Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/os.rs:146]: The std docs don't makey any such promise though: https://github.com/rust-lang/rust/blob/9649706eada1b2c68cf6504356efb058f68ad739/library/std/src/env.rs#L58
In fact, it seems fine if there are some platform specific behaviour here.

UEFI has no notion of parent or child. It is like a single-tasking os.

**Comment by Ayush1325** [comment]
ping @joboet

**Comment by joboet** [comment]
I'm very sorry for not getting back to you sooner.

The problem I see here is that `GetCurDir` returns a mapping, not a device path and will therefore not interoperate with our current filesystem implementation. I think there are two options we could take here:
1. Normalize the mapping to a device path using `GetDevicePathFromMap`
2. Use the filesystem functions from UEFI shell when available, which support both mappings and device paths.

Option 2 is more work, but has the advantage that `File` would support opening paths stored in environment variables, which will probably be mappings, not device paths, so I think that is the best option in the long run. If you want to get this PR merged first, then option 1 would also be fine for now.

**Comment by joboet** [comment]
Right, then let's ship this.
@bors r+

**Comment by matthiaskrgr** [comment]
@bors r- 
https://github.com/rust-lang/rust/pull/131665#issuecomment-2409817167

**Comment by Ayush1325** [comment]
@joboet @matthiaskrgr I think this should be ready to merge now.

**Pull request merged**